custom_components/google_keep_sync/__init__.py

```diff
@@ -28,13 +30,72 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
         _LOGGER.error("Failed to authenticate Google Keep API")
         return False  # Exit early if authentication fails
 
+    async def _parse_gkeep_data_dict() -> dict[str, dict[str, dict[str, str]]]:
+        """Parse unchecked gkeep api data to a dictionary."""
+        todo_lists = {}
+
+        # for each list
+        for glist in coordinator.data or []:
+            # get all the unchecked items only
+            items = {
+                item.id: {"summary": item.text, "checked": item.checked}
+                for item in glist.items
+                if not item.checked
+            }
+            todo_lists[glist.id] = {"name": glist.title, "items": items}
+        return todo_lists
+
+    async def _check_gkeep_lists_changes(original_lists, updated_lists) -> None:
+        """Compare original_lists and updated_lists lists.
+
+        Report on any new TodoItem's that have been added to any
+        lists in updated_lists that are not in original_lists.
+        """
+        # for each list
+        for upldated_list_id, upldated_list in updated_lists.items():
+            if upldated_list_id not in original_lists:
+                _LOGGER.debug(
+                    "Found new list not in original: %s", upldated_list["name"]
+                )
+                continue
+
+            # for each todo item in the list
+            for upldated_list_item_id, upldated_list_item in upldated_list[
+                "items"
+            ].items():
+                # if todo is not in original list, then it is new
+                if (
+                    upldated_list_item_id
+                    not in original_lists[upldated_list_id]["items"]
+                ):
+                    list_prefix = entry.data.get("list_prefix", "")
+                    data = {
+                        "item_name": upldated_list_item["summary"],
+                        "item_id": upldated_list_item_id,
+                        "item_checked": upldated_list_item["checked"],
+                        "list_name": (f"{list_prefix} " if list_prefix else "")
+                        + upldated_list["name"],
+                        "list_id": upldated_list_id,
+                    }
+
+                    _LOGGER.debug("Found new TodoItem: %s", data)
+                    hass.bus.async_fire("google_keep_sync_new_item", data)
```

Comment:
When I first read the function my initial thought from the function name was that we were just checking for changes, it didn't occur to me that we would be firing an event from the function. The fact that there is a side effect might warrant renaming the function so it's clear we're doing something with the changes. Might also be worth it to just use a callback instead of directly calling hass.bus.async_fire here (that way we don't have to pass the hass parameter and the function is more generic/adaptable).

Here is my suggestion:
```
from collections import namedtuple
from typing import Callable

TodoItemData = namedtuple(
    "TodoItemData", ["item_name", "item_id", "item_checked", "list_name", "list_id"]
)

async def _handle_new_items_added(
    original_lists: dict[str, dict],
    updated_lists: dict[str, dict],
    list_prefix: str,
    on_new_item: Callable[[TodoItemData], None],
) -> None:
    """Compare original and updated lists to find new TodoItems.

    For each new TodoItem found, call the provided on_new_item callback.

    :param original_lists: The original todo list data.
    :param updated_lists: The updated todo list data.
    :param on_new_item: Callback function to execute for each new item found.
    """
    for updated_list_id, updated_list in updated_lists.items():
        if updated_list_id not in original_lists:
            _LOGGER.debug("Found new list not in original: %s", updated_list["name"])
            continue

        for updated_list_item_id, updated_list_item in updated_list["items"].items():
            if updated_list_item_id not in original_lists[updated_list_id]["items"]:
                item_data = TodoItemData(
                    item_name=updated_list_item["summary"],
                    item_id=updated_list_item_id,
                    item_checked=updated_list_item["checked"],
                    list_name=(f"{list_prefix} " if list_prefix else "")
                    + updated_list["name"],
                    list_id=updated_list_id,
                )
                _LOGGER.debug("Found new TodoItem: %s", item_data)
                await on_new_item(item_data)
```

Then in async_update_data:
```
result = await api.async_sync_data(lists_to_sync)
# save lists after syncing
updated_lists = await _parse_gkeep_data_dict()
# compare both list for changes
await _handle_new_items_added(
    original_lists,
    updated_lists,
    entry.data.get("list_prefix", ""),
    lambda item_data: hass.bus.async_fire(
        "google_keep_sync_new_item", item_data._asdict()
    ),
)
```

Reply:
With this, I'm curious why you suggested a callable vs just returning a list of new items which could be iterated over?

Reply:
That's a good solution too. So something like 
```
new_items = await _get_new_todo_items(original_lists, updated_lists, entry.data.get("list_prefix", ""))

for new_item in new_items:
    await hass.bus.async_fire("google_keep_sync_new_item", new_item._asdict())
```

I think I was just thinking in terms of conciseness since we were already iterating within the existing function. But honestly either way is fine so feel free to change this. I just wanted to get the hass.bus.async_fire out of that function so we had a single concern/responsibility for that function.